auth/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The `google-cloud-auth` crate provides a convenient way of fetch credentials
+//! from various environments. This process is of finding credentials from the
+//! environment is called [Application Default Credentials](https://google.aip.dev/auth/4110).
+
+use chrono::Utc;
+use chrono::{DateTime, Duration};
+use serde::Deserialize;
+use source::*;
+use std::path::PathBuf;
+
+mod metadata;
+mod oauth2;
+mod source;
+
+const GOOGLE_APPLICATION_CREDENTIALS_ENV: &str = "GOOGLE_APPLICATION_CREDENTIALS";
+const WINDOWS_APPDATA_ENV: &str = "APPDATA";
+const UNIX_HOME_ENV: &str = "HOME";
+const USER_CREDENTIAL_FILE: &str = "application_default_credentials.json";
+const GCLOUD_PATH_PART: &str = "gcloud";
+const CONFIG_PATH_PART: &str = ".config";
+
+#[derive(thiserror::Error, Debug)]
+#[non_exhaustive]
+pub enum Error {
+    #[error("unable to read file")]
+    Io(#[from] std::io::Error),
+    #[error("unable to deserialize value")]
+    Serde(#[from] serde_json::Error),
+    #[error("unable to process request")]
+    Http(#[from] reqwest::Error),
+    #[error("{0}")]
+    Other(String),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// AccessToken holds a token value that can be used in Authorization headers to
+/// authenticate with Google Cloud APIs.
+#[derive(Clone)]
+#[non_exhaustive]
+pub struct AccessToken {
+    /// The actual token.
+    pub value: String,
+    expires: Option<DateTime<Utc>>,
+}
+
+impl AccessToken {
+    /// Returns true if the token should be considered valid, compensating for
+    /// clock skew with by ten seconds.
+    pub(crate) fn is_validish(&self) -> bool {
+        if let Some(expires) = self.expires {
+            let now = Utc::now();
+            // Avoid clock skew with 10 second diff of now.
+            let expiresish = expires - Duration::seconds(10);
+            expiresish > now
+        } else {
+            false
+        }
+    }
+}
+
+/// Configuration for various authentication flows.
+#[non_exhaustive]
+pub struct CredentialConfig {
```

Comment:
I now concur we should use builders because:
1. We can't add public members to structs without taking a major version number upgrade.
2. Having a separate builder type with a `.build()` method means we validate the parameters exactly once at build time, instead of validating it at the beginning of each function call.

Reply:
Done.